src/box/rt_access.c

```diff
@@ -0,0 +1,158 @@
+#include "rt_access.h"
+#include "assoc.h"
+#include "../lua/init.h" /* tarantool_lua_is_builtin_global */
+#include "salad/grp_alloc.h"
+
+static struct mh_strn_strn_t *user_rt_access = NULL;
+static struct mh_strnptr_t *user_rt_universe_access = NULL;
+
+struct mh_strnptr_key_t*
+mh_strnptr_key_alloc(const char *str, uint32_t str_len)
+{
+	struct mh_strnptr_key_t *node;
+	struct grp_alloc all = grp_alloc_initializer();
+	grp_alloc_reserve_data(&all, sizeof(*node));
+	grp_alloc_reserve_str(&all, str_len);
+	grp_alloc_use(&all, xmalloc(grp_alloc_size(&all)));
+	void *p = grp_alloc_create_data(&all, sizeof(*node));
+	node = (struct mh_strnptr_key_t *)p;
+	node->str = grp_alloc_create_str(&all, str, str_len);
+	node->len = str_len;
+	assert(grp_alloc_size(&all) == 0);
+	return node;
+}
+
+struct mh_strn_strn_key_t*
+mh_strn_strn_key_alloc(const char *s1, uint32_t s1_len,
+		       const char *s2, uint32_t s2_len)
+{
+	struct mh_strn_strn_key_t *node;
+	struct grp_alloc all = grp_alloc_initializer();
+	grp_alloc_reserve_data(&all, sizeof(*node));
+	grp_alloc_reserve_str(&all, s1_len);
+	grp_alloc_reserve_str(&all, s2_len);
+	grp_alloc_use(&all, xmalloc(grp_alloc_size(&all)));
+	void *p = grp_alloc_create_data(&all, sizeof(*node));
+	node = (struct mh_strn_strn_key_t *)p;
+	node->s1 = grp_alloc_create_str(&all, s1, s1_len);
+	node->s1_len = s1_len;
+	node->s2 = grp_alloc_create_str(&all, s2, s2_len);
+	node->s2_len = s2_len;
+	assert(grp_alloc_size(&all) == 0);
+	return node;
+}
+
+void
+mh_strnptr_free(struct mh_strnptr_t *h)
+{
+	if (h == NULL)
+		return;
+
+	mh_int_t i;
+	mh_foreach(h, i) {
+		struct mh_strnptr_node_t *node = mh_strnptr_node(h, i);
+		if (node != NULL)
+			free(node->val);
+	}
```

Comment:
`free()` must accept the NULL pointer according to the [POSIX standard](https://pubs.opengroup.org/onlinepubs/007904975/functions/free.html).

So, we can reduce this code snippet down to:

```c
mh_foreach(h, i)
	free(mh_strnptr_node(h, i)->val);
```

(The same for `mh_strn_strn_free` below.)

Reply:
Done.